File: 0016AddTwoNumbersAgain/2-AddTwoNumbers.java

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummyHead = new ListNode(0);
    ListNode current = dummyHead;
    int carry = 0;
    
    while (l1 != null || l2 != null) {
        int value1 = (l1 != null) ? l1.val : 0;
        int value2 = (l2 != null) ? l2.val : 0;
        int sum = value1 + value2 + carry;
        carry = sum / 10;
        int digit = sum % 10;
        
        current.next = new ListNode(digit);
        current = current.next;
        
        l1 = (l1 != null) ? l1.next : null;
        l2 = (l2 != null) ? l2.next : null;
    }
    
    if (carry > 0) {
        current.next = new ListNode(carry);
    }
    
    return dummyHead.next;
    }
}

File: 2-add-two-numbers/2 add two numbers.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    const result = new ListNode();
    let crawlerResult = result;
    let lastCrawler: ListNode;
    let crawlerL1 = l1;
    let crawlerL2 = l2;
    let carry = 0;

    while (crawlerL1 || crawlerL2) {
        crawlerResult.next = new ListNode();

        if (crawlerL1 && crawlerL2) {
            const sum = crawlerL1.val + crawlerL2.val + carry;
            crawlerResult.val = sum % 10;
            carry = Math.floor(sum / 10);
        } else {
            const crawler = crawlerL1 ? crawlerL1 : crawlerL2;

            const sum = crawler.val + carry;
            crawlerResult.val = sum % 10;
            carry = Math.floor(sum / 10);
        }

        crawlerL1 = crawlerL1?.next;
        crawlerL2 = crawlerL2?.next;
        lastCrawler = crawlerResult;
        crawlerResult = crawlerResult.next;
    }

    if (carry === 1) {
        crawlerResult.val = carry;
    } else {
        lastCrawler.next = null;
    }

    return result;
};